Lumos 6.23 AKS walkthrough (#4)
CrowdStrike/Cloud-Azure: container/falcon-container-aks-implementation-guide.md

# Implementation Guide for CrowdStrike Falcon-Container Sensor on Azure AKS Kubernetes cluster

> This guide works through creation of a new Kubernetes cluster, deployment of the Falcon-Container Sensor, and demonstration of detection capabilities of Falcon Container Workload Protection.
> 

Reviewer: ```suggestion
This guide works through the creation of a new Azure Kubernetes Service (AKS) cluster, deployment of the Falcon Container Sensor, and demonstration of detection capabilities of Falcon Container Workload Protection.

```

> No prior Kubernetes or Falcon knowledge is needed to follow this guide. First sections of this guide focus on creation of ACR Container registry and an AKS cluster, however, these sections may be skipped if you have access to an existing cluster.

Reviewer: ```suggestion
No prior Kubernetes or Falcon knowledge is needed to follow this guide. The first sections of this guide focus on the creation of an Azure Container Registry (ACR) and an AKS cluster, however, these sections may be skipped if you have access to an existing registry and cluster.
```

> Time needed to follow this guide: 45 minutes.

Reviewer: Nice call setting the time expectation upfront!

## Pre-requisites

> - Refer to the lastest documentation on the [falcon-container](https://falcon.crowdstrike.com/documentation/146/falcon-container-sensor-for-linux)

Reviewer: For what purpose? There's a lot of text there. Anything in particular they should review ahead of time?

- Existing Azure Subscription and Global Adminstrator permissions
- You will need a workstation to complete the installation steps below
  * These steps have been tested on Linux and should also work with OSX
- Install [docker](https://www.docker.com/products/docker-desktop) container runtime

> - Verify docker daemon is running on your workstation

Reviewer: Would it improve clarity by saying something like "A workstation or virtual machine running the Docker container runtime" ?

- API Credentials from Falcon with Sensor Download Permissions

>   * These credentials can be created in the Falcon platform under Support->API Clients and Keys.

Reviewer: Can we provide a link?

  * For this step and practice of least privilege, you would want to create a dedicated API secret and key.
- Azure Cli installed locally and authenticated

```
az login
```
- Install kubectl
```
az aks install-cli

> ```

Reviewer: Where can Azure CLI be downloaded from?

## Deployment

### Step 1: Setup an Azure Container Registry

- Set your ACR registry name and resource group name into variables
- Note: The ACR_NAME must be a unique name globally as a DNS record is created to reference the image registry
```
CLOUD_REGION=westus
ACR_NAME=<arc_unique_name>
RG_NAME=rg_cswest

> ```

Reviewer: should those values for CLOUD_REGION and RG_NAME be copy/pasted, or should users change them? If changed, to what values?

- Create the resource group for the ACR and Cluster
```
az group create --name $RG_NAME --location $CLOUD_REGION
```
- Create the Azure Container Registry
```
az acr create --name $ACR_NAME --sku basic -g $RG_NAME --location $CLOUD_REGION
```
  Example output:
```
    [
    {
        "adminUserEnabled": false,
        "anonymousPullEnabled": false,
        "creationDate": "2021-04-16T21:12:03.724818+00:00",
        "dataEndpointEnabled": false,
        "dataEndpointHostNames": [],
        "encryption": {
        "keyVaultProperties": null,
        "status": "disabled"
        },

>         "id": "/subscriptions/9da62014-f7c1-4e8f-9241-14a2b28362f1/resourceGroups/rg_cswest/providers/Microsoft.ContainerRegistry/registries/csDemoAcr01",

Reviewer: Do we care if subscription IDs are exposed? Genuinely do not know enough about Azure to make a call on this. Generally, AWS Account IDs are kept secret.

Author: Hmm... AWS Account IDs do not change but I could easily kill this subscription and create another in the tenant. Since this is already in the history I will probably just do that and leave it out in future  output. Good catch.

        "identity": null,
        "location": "westus",
        "loginServer": "csdemoacr01.azurecr.io",
        "name": "csDemoAcr01",
        "networkRuleBypassOptions": "AzureServices",
        "networkRuleSet": null,
        "policies": {
        ....
```

> - Notate the login server from the output shown above and add to a new variable

Reviewer: ```suggestion
- Notate the ``loginServer`` value from the output shown above and add to a new environment variable
```

```
ACR_LOGINSERVER=csdemoacr01.azurecr.io
```

### Step 2: Download the falcon-container sensor

Note: For existing ACR set the loginServer attribute from the ACR to the following variable.
```
ACR_LOGINSERVER=<loginServer>
```
- Set the FALCON_IMAGE_UI variable to the full path of the tagged image and repository for future use
```

> FALCON_IMAGE_URI=$ACR_LOGINSERVER/falcon-sensor:latest

Reviewer: Will the image always be called ``falcon-sensor``? Where was this specified?

Author: This is the repo name we are giving it in the retag so it is arbitrary really. falco-sensor is the name product gives it so we just carry it along. The manifest will use this later on in the -image flag so we could probably call it whatever we want. This naming should work for almost all cases unless we had multiple falcon-sensor versions we were using. This should be good for this walkthrough.

```

> - Login to the Falcon Platform and Navigate to Hosts->Sensor Downloads

Reviewer: Can we provide links? eg for falcon.crowdstrike.com and falcon.us-2.crowdstrike (or whatever it is?)

> - Download the latest 'Container' sensor (Support for AKS was added in 6.23.601)

Reviewer: Suggest clarifying that 6.23.601+ will work, else it may be assumed people need EXACTLY 6.23.601

- Import the tarball to your local docker. If you are following this guide inside the tooling container, you can run this command outside of the container as the docker socket is shared between your host system and the said tooling container.
```
docker load -i falcon-sensor-6.23.0-601.container.x86_64.tar.bz2
```
Example output:
```
ad2fea6dc0f1: Loading layer [==================================================>]   78.6MB/78.6MB
Loaded image: falcon-sensor:6.23.0-601.container.x86_64.Release.US1
```
- List the loaded docker image
```
docker images | grep falcon-sensor
```
Example output:
```
   falcon-sensor  6.23.0-601.container.x86_64.Release.US1   e618c8efcc93   2 weeks ago    78.5MB
```

> - Note the image name and the image tag in the first and second columns respectively

Author: The grep removed it.. I changed the grep to add the REPOSITORY expression and updated the output.

- Using the local image name and tag, re-tag the image for your managed ECR using the variable previously set
```
docker tag falcon-sensor:6.23.0-601.container.x86_64.Release.US1 $FALCON_IMAGE_URI
```
- Authenticate to the ACR
```
 az acr login -n $ACR_NAME
```
Example output:
```
    Uppercase characters are detected in the registry name. When using its server url in docker commands, to avoid authentication errors, use all lowercase.
    Login Succeeded
```
- Upload the falcon-container-sensor to the ACR you created previously
```
docker push $FALCON_IMAGE_URI
```

### Step 3: Create the AKS cluster

- Set the name of the AKS Cluster into a variable
```
AKS_CLUSTER=csAksCluster01
```
- Create the AKS Cluster and attach the ACR
```

> az aks create --name $AKS_CLUSTER --kubernetes-version 1.18.14 --attach-acr $ACR_NAME -g $RG_NAME --generate-ssh-keys

Author: Looks like this should work on any AKS with k8s >=v1.18 (for now). I will add the tested on phrases for clarification anyhow.
